ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java

```diff
@@ -375,6 +382,76 @@ protected void onAttachedToWindow() {
   }
 
   @Override
+  public void addView(View child, int index, LayoutParams params) {
+    // This will get called for every overload of addView so there is not need to override every method.
+    if (ReactViewManager.getViewZIndex(child) != null) {
+      mNumberOfChildrenWithZIndex++;
+    }
+
+    setChildrenDrawingOrderEnabled(mNumberOfChildrenWithZIndex > 0);
+    mDrawingOrderIndices = null;
+
+    super.addView(child, index, params);
+  }
+
+  @Override
+  public void removeView(View view) {
+    handleRemoveView(view);
+
+    super.removeView(view);
+  }
+
+  @Override
+  public void removeViewAt(int index) {
+    handleRemoveView(getChildAt(index));
+
+    super.removeViewAt(index);
+  }
+
+  private void handleRemoveView(View view) {
+    if (ReactViewManager.getViewZIndex(view) != null) {
+      mNumberOfChildrenWithZIndex--;
+    }
+
+    setChildrenDrawingOrderEnabled(mNumberOfChildrenWithZIndex > 0);
+    mDrawingOrderIndices = null;
+  }
+
+  @Override
+  protected int getChildDrawingOrder(int childCount, int index) {
```

Comment:
one issue with this approach is that getChildDrawingOrder needs to be implemented by each ViewGroup (ScrollView, ModalHostView, any custom ViewGroups) in order to support z index for that specific ViewGroup. 

1. can you add a warning for usages of z index if the parent ViewGroup doesn't support z index? possibly by using an interface
2. can you extract the add, remove, child drawing order logic into a separate class to make it easier to reuse this logic across other ViewGroups if needed? a unit test for this new class would be useful as well